packages/block-library/src/heading/heading-checker.js

```diff
@@ -0,0 +1,58 @@
+/**
+ * WordPress dependencies
+ */
+import { speak } from '@wordpress/a11y';
+import { __ } from '@wordpress/i18n';
+import { Notice } from '@wordpress/components';
+import { useEffect } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
+import { withSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { computeOutlineHeadings } from '../../../editor/src/components/document-outline/index.js';
+
+export const HeadingChecker = ( { blocks = [], selectedLevel, selectedHeadingId } ) => {
+	const headings = computeOutlineHeadings( blocks );
+
+	// Iterate headings to find prevHeadingLevel
+	let prevHeadingLevel = 1;
+	let i = 0;
+	for ( i = 0; i < headings.length; i++ ) {
+		if ( i >= 1 && headings[ i ].clientId === selectedHeadingId ) {
+			prevHeadingLevel = headings[ i - 1 ].level;
+		}
+	}
+
+	const isIncorrectLevel = ( selectedLevel === 1 || selectedLevel > prevHeadingLevel + 1 );
+
+	if ( ! isIncorrectLevel ) {
+		return null;
+	}
+
+	const msg = __( 'This heading level is incorrect. ' );
+
+	// For accessibility
+	useEffect( () => {
+		speak( __( 'This heading level is incorrect' ) );
+	}, [ selectedLevel ] );
+
+	return (
+		<div className="editor-heading-checker block-editor-heading-checker">
+			<Notice status="warning" isDismissible={ false }>
+				{ msg }
+			</Notice>
+		</div>
+	);
+};
+
+export default compose(
+	withSelect( ( select ) => {
+		const { getBlocks } = select( 'core/block-editor' );
```

Comment:
@jorgefilipecosta - is there a way to get all heading blocks only so we didn't have to use `computeOutlineHeadings` in the first place?

There is [getGlobalBlockCount](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/store/selectors.js#L270-L285), so something similar would be nice.

That way we could share the same selector in both places and remove this `computeOutlineHeadings` helper altogether.

Reply:
In fact, `headings` has been computed in 
https://github.com/WordPress/gutenberg/blob/8722f49d3614a9c71d7a6a3af06eebec4c9bfa5b/packages/editor/src/components/document-outline/index.js#L67-L69
I am not sure whether we could avoid computing it again.

Reply:
We can if we create memoized shared selector, similar to how `getBlocks` is cached as of today. It might be a complicated task and that's why I asked @jorgefilipecosta for help.

Reply:
We could create a generic selector that returns the blocks of a given type, so it is not specific to this functionality.
But I'm not sure if it would fit our needs, computeOutlineHeadings does more than just returning the heading blocks, it computes, for example, the heading path required to show the path where a heading is nested on.
I think in this specific case just the selector would work, but in the checks, we do in the Content Structure we would still need something like computeOutlineHeadings to associate a path to each heading.
Exposing a selector that does what computeOutlineHeadings does (or similar functionality with paths) seems very specific to this functionality.
I guess a possible path here is just implementing level-checker.js as part of the editor module and reusing computeOutlineHeadings there.